### PR TITLE
[tools] handle commits merged without PR in deprecated script

### PR DESCRIPTION
**Pull request by asraa**
Some commits (eg https://github.com/envoyproxy/envoy/commit/f5b0294f0a7c2d143f6d4d94afae9a0e6f9dacf7 ) were added without a PR, so the script fails. Change so that issues can still be created. 

This uses the commit message and sha instead of the PR title and number.

Testing:
I successfully got: 
`Flag envoy.reloadable_features.http1_flood_protection added at 2020-03-02 is not ready to remove`
Instead of a crash. Tweaking the script to deprecate after 1 day and printing out the issue and title give me:
```
envoy.reloadable_features.http1_flood_protection deprecation
commit f5b0294f0a7c2d143f6d4d94afae9a0e6f9dacf7 (http: adding response flood protection) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to alyssawilk
```

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

File: tools/deprecate_version/deprecate_version.py [hunk 1]
```diff
@@ -90,47 +90,61 @@ def CreateIssues(access_token, runtime_and_pr):
...
+      pr_info = repo.get_pull(pr)
+      change_title = pr_info.title
+      number = ('#%d') % pr
+      login = pr_info.user.login
+    else:
+      # Extract Commit message, sha, and author.
+      # In this case we cannot add a user login for an assignee.
+      change_title = commit.message.split('\n')[0] # Remove sign-off mesage
+      number = ('commit %s') % commit.hexsha
+      login = None
```
Comment (asraa) [inline tools/deprecate_version/deprecate_version.py:106]: GitPython API can only give me author name and email, so I could not assign the issue to the NamedUser login like before. GitPython experts, is there a way?

File: tools/deprecate_version/deprecate_version.py [hunk 2]
```diff
@@ -90,47 +90,61 @@ def CreateIssues(access_token, runtime_and_pr):
...
     # Who is the author?
-    pr_info = repo.get_pull(pr)
+    if pr:
+      # Extract PR title, number, and author.
+      pr_info = repo.get_pull(pr)
+      change_title = pr_info.title
+      number = ('#%d') % pr
+      login = pr_info.user.login
+    else:
+      # Extract Commit message, sha, and author.
```
Comment (alyssawilk) [inline tools/deprecate_version/deprecate_version.py:102]: Commit - commit

File: tools/deprecate_version/deprecate_version.py [hunk 3]
```diff
@@ -90,47 +90,61 @@ def CreateIssues(access_token, runtime_and_pr):
...
+      # In this case we cannot add a user login for an assignee.
+      change_title = commit.message.split('\n')[0]  # Remove sign-off mesage
```
Comment (alyssawilk) [inline tools/deprecate_version/deprecate_version.py:104]: It's not sign off but description too (if there is one) yes?

**Comment by asraa** [comment]
Output when running currently:

```
Flag envoy.reloadable_features.http1_flood_protection added at 2020-03-02 is not ready to remove
Flag envoy.reloadable_features.strict_header_validation added at 2019-11-13 is not ready to remove
Flag envoy.reloadable_features.connection_header_sanitization added at 2019-11-18 is not ready to remove
Flag envoy.reloadable_features.strict_authority_validation added at 2019-12-10 is not ready to remove
Flag envoy.reloadable_features.reject_unsupported_transfer_encodings added at 2019-12-13 is not ready to remove
Flag envoy.reloadable_features.strict_method_validation added at 2020-01-09 is not ready to remove
Flag envoy.reloadable_features.new_http1_connection_pool_behavior added at 2020-01-29 is not ready to remove
Flag envoy.reloadable_features.new_http2_connection_pool_behavior added at 2020-01-29 is not ready to remove
Flag envoy.reloadable_features.ext_authz_http_service_enable_case_sensitive_string_matcher added at 2020-02-25 is not ready to remove
Found end sentinel

No code is deprecated.
```

If I change the script so I deprecate after 1 Day I get:
```
Flag envoy.reloadable_features.http1_flood_protection added at 2020-03-02 and is safe to remove
Flag envoy.reloadable_features.strict_header_validation added at 2019-11-13 and is safe to remove
Flag envoy.reloadable_features.connection_header_sanitization added at 2019-11-18 and is safe to remove
Flag envoy.reloadable_features.strict_authority_validation added at 2019-12-10 and is safe to remove
Flag envoy.reloadable_features.reject_unsupported_transfer_encodings added at 2019-12-13 and is safe to remove
Flag envoy.reloadable_features.strict_method_validation added at 2020-01-09 and is safe to remove
Flag envoy.reloadable_features.new_http1_connection_pool_behavior added at 2020-01-29 and is safe to remove
Flag envoy.reloadable_features.new_http2_connection_pool_behavior added at 2020-01-29 and is safe to remove
Flag envoy.reloadable_features.ext_authz_http_service_enable_case_sensitive_string_matcher added at 2020-02-25 and is safe to remove
Found end sentinel

envoy.reloadable_features.http1_flood_protection deprecation
commit f5b0294f0a7c2d143f6d4d94afae9a0e6f9dacf7 (http: adding response flood protection) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to alyssar@chromium.org
  >> Issue already exists, not posting!
envoy.reloadable_features.strict_header_validation deprecation
#8849 (release: flipping flags) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to alyssawilk
envoy.reloadable_features.connection_header_sanitization deprecation
#8862 (Sanitize headers nominated by the Connection header) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to abaptiste
envoy.reloadable_features.strict_authority_validation deprecation
#77 (docs: small fix) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to mattklein123
envoy.reloadable_features.reject_unsupported_transfer_encodings deprecation
#9316 (http: disallow unknown transfer encodings) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to alyssawilk
envoy.reloadable_features.strict_method_validation deprecation
#9637 (http: explicitly blocking CONNECT for HTTP/1.1) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to alyssawilk
envoy.reloadable_features.new_http1_connection_pool_behavior deprecation
#9668 (Combine H1 and H2 conn_pool implementations) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to ggreenway
envoy.reloadable_features.new_http2_connection_pool_behavior deprecation
#9668 (Combine H1 and H2 conn_pool implementations) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to ggreenway
envoy.reloadable_features.ext_authz_http_service_enable_case_sensitive_string_matcher deprecation
#9715 (ext_authz: Deprecate LowerCaseStringMatcher) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to dio
```

File: tools/deprecate_version/deprecate_version.py (hunk 1 above)
Comment (alyssawilk) [reply tools/deprecate_version/deprecate_version.py:106]: Not that I know of.  Does assigning to email work?  Worst case we could assign the script runner to find an owner and print that's what is going on, but it'd be nice if we could do better.

Comment (asraa) [reply tools/deprecate_version/deprecate_version.py]: I found `search_users` functionality in the API -- I did guard against not finding the user, but I think (?) there's guarantee there will be a user with the email we find.

File: tools/deprecate_version/deprecate_version.py (hunk 2 above)
Comment (asraa) [reply tools/deprecate_version/deprecate_version.py]: Done

File: tools/deprecate_version/deprecate_version.py (hunk 3 above)
Comment (asraa) [reply tools/deprecate_version/deprecate_version.py]: Yep -- that's right it.

File: tools/deprecate_version/deprecate_version.py (hunk 1 above)
Comment (asraa) [reply tools/deprecate_version/deprecate_version.py]: Edit: I trial ran it without creating the issue, I successfully got your login with the script now:
```
envoy.reloadable_features.http1_flood_protection deprecation
commit f5b0294f0a7c2d143f6d4d94afae9a0e6f9dacf7 (http: adding response flood protection) introduced a runtime guarded feature. This issue tracks source code cleanup.
  >> Assigning to alyssawilk
```

**Comment by mattklein123** [comment]
Please merge master. Also @junr03 PTAL. Thank you!

/wait

File: tools/deprecate_version/deprecate_version.py [hunk 4]
```diff
@@ -70,47 +71,62 @@ def CreateIssues(access_token, runtime_and_pr):
...
+    if pr:
+      # Extract PR title, number, and author.
+      pr_info = repo.get_pull(pr)
+      change_title = pr_info.title
+      number = ('#%d') % pr
+      login = pr_info.user.login
+    else:
+      # Extract commit message, sha, and author.
+      # In this case we cannot add a user login for an assignee.
+      change_title = commit.message.split('\n')[0]  # Remove description.
```
[the remark this answers is not in the document]
Comment (asraa) [reply tools/deprecate_version/deprecate_version.py]: Oh maybe, it looks like there's no enforced limit on the commit title. People seem to recommend 50/72 stylistically? I arbitrarily put 50.

File: tools/deprecate_version/deprecate_version.py [hunk 5]
```diff
@@ -140,14 +156,15 @@ def GetRuntimeAndPr():
         if runtime_guard == 'envoy.reloadable_features.test_feature_true':
           found_test_feature_true = True
           continue
-        pr = (int(re.search('\(#(\d+)\)', commit.message).group(1)))
+        pr_num = re.search('\(#(\d+)\)', commit.message)
+        pr = (int(pr_num.group(1))) if pr_num else None
```
[the remark this answers is not in the document]
Comment (asraa) [reply tools/deprecate_version/deprecate_version.py:162]: All set

**Pull request merged**